### PR TITLE
Use widing refactoring to allow passing a lazy evaluated CharSequence (#1133)

### PR DESCRIPTION
**Pull request by ruckc**
This allows for using the included `Message.message()` to generate lazy evaluated CharSequence based on String.format

### CODE REVIEW

**Comment by nathantippy** [comment]
This is expected to be a large change but one with little risk.  All Strings are instances of CharSequence.  The message parameter is of type String and only used when the test fails.  If message were a CharSequence the existing behavior would not change.   

The benefit of this change would be to enable some classes that extend CharSquence to permit lazy construction of the message.  In this way expensive detail gathering could be one only upon failure, eg use of the message.

File: src/main/java/org/junit/Message.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    public String toString() {
+        evaluate();
+        return evaluated;
+    }
+    
+    public static Message message(String format, Object ... arguments) {
+        return new Message(format, arguments);
+    }
+    
+    public static Message msg(String format, Object ... arguments) {
```
Comment (ashleyfrieze) [inline src/main/java/org/junit/Message.java]: Why have both a msg and a message function? Is this duplication a real benefit?

**Comment by ashleyfrieze** [comment]
I like the idea of this, but I'm not sure about the reality. A couple of observations. One, any third party extensions to JUnit will continue to use Strings for messages, rather than CharSequence, which may lead to people having to blend between Message and String. Two: the painful effort in initialisation of strings in real life, is usually in retrieving what to put into the string, not making the string itself. That won't be removed by the Message pattern.

E.g. message("Something: %s", something.toString()); // the effort is in compiling the output of something.toString, not in compiling the message - something.toString() might require a lot of fields to be examined by reflection or something. That would be evaluated when creating the Message object, even though it doesn't run through String.format if it's not used.

If this could be used with lambdas, then it would make a difference.

e.g. message("Something: %s", something -> something.toString());

In this situation, the lambda would not be evaluated at the point of creating Message and would genuinely save time if the message were never needed.

**Comment by kcooney** [comment]
@ashleyfrieze you could postpone producing the value to be put into the string by passing in an Object as a parameter, and producing the value in the object's `toString()`, but agreed that the lambda syntax is cleaner

**Comment by ashleyfrieze** [comment]
@kcooney love your thinking. But practically it would be this...

From:

    new Message("Broken %s", myBigThing.getBigData());

to

    new Message("Broken %S", new Object() { @Override public void toString() { myBigThing.getBigData()); }});

That's quite a workaround :)

**Comment by ashleyfrieze** [comment]
Perhaps there's an alternative to this by adding an overload to assertThat allowing for a just-in-time message creation.

    public interface Message {
        String getMessage();
    }


    public void assertThat(Message message, Object o, Matcher m) {
        ...
        // pseudo code
        fail(message.getMessage);
     }

called thusly:

    assertThat(new Message() { public String getMessage() { return myObject.getBigData(); }},
        someValue, is(expected));

The benefit of changing just assertThat is that it's a smaller change. The benefit of the Message interface is that it could easily be subclassed by the user for reusable messages. We could also provide a Java 8 message provider which uses lambda:

    assertThat(new LambdaMessage(() -> myObject.getBigData()), someValue, is(expected));

Just tossing the ideas out here.

**Comment by marcphilipp** [comment]
Actually, the equivalent lambda expression would be:

```java
assertThat(() -> myObject.getBigData(), someValue, is(expected));
```

**Comment by marcphilipp** [comment]
However, we want to get rid of the Hamcrest dependency in the future. Can you try to submit a similar pull request there for [`MatcherAssert`](https://github.com/hamcrest/JavaHamcrest/blob/master/hamcrest-core/src/main/java/org/hamcrest/MatcherAssert.java)?

**Comment by ashleyfrieze** [comment]
@marcphilipp my suggestion was for a Java 8 adapter over a Java 5 interface. If everything were Java 8 this would all be moot!

**Comment by marcphilipp** [comment]
I understand. However, for `assertThat` I think you should ask over at hamcrest/JavaHamcrest.

**Comment by ashleyfrieze** [comment]
See https://github.com/hamcrest/JavaHamcrest/pull/102 - which is this turned into a potential hamcrest change.

**Comment by kcooney** [comment]
I'm closing this since mark closed 1133

**Pull request closed without merging**